File: fsharp/queen-attack/QueenAttack.fs
module QueenAttack

let create ((r, c): int * int) = 
    r >= 0 && r <= 7 && c >= 0 && c <= 7

let canAttack (queen1: int * int) (queen2: int * int) = 
    let (r1, c1) = queen1
    let (r2, c2) = queen2
    r1 = r2 || c1 = c2 || abs (r2 - r1) = abs (c2 - c1)

File: c/queen-attack/1/queen_attack.c
#include "queen_attack.h"
#include <stdlib.h>
#include <stdbool.h>

attack_status_t can_attack(position_t white_queen, position_t black_queen) {
    if (white_queen.row > 7 || black_queen.row > 7 ||
        white_queen.column > 7 || black_queen.column > 7 ||
        (white_queen.row == black_queen.row && white_queen.column == black_queen.column)
    ) return INVALID_POSITION;

    if (white_queen.row == black_queen.row ||
        white_queen.column == black_queen.column ||
        abs(white_queen.row - black_queen.row) == abs(white_queen.column - black_queen.column) 
    ) return CAN_ATTACK;
    
    return CAN_NOT_ATTACK;
}
